### ISSUE TITLE
Issue with jax.lax.scan and jit

### ISSUE THREAD

**Issue by stefansommer** [google/jax#7531]
Dear JAX team,

thank for your excellent library!

As far as I can tell, the code below shows an issue with jax.lax.scan and jit. lax.scan gives result different compared to without jit, compared to a reference python scan from the scan documentation and a direct application of the function the scan runs over. Moreover, making changes to the code that should not affect the result changes the output of scan (see comments in code). The code is extracted from a numerical ODE integration which gave erroneous results. The code therefore does not make sense as such, but it illustrates the issue. The function f to be scanned over can perhaps be further simplified, but as you may notice making even trivial changes to it (e.g. changing values that are multiplied by 0) changes the output. It seems to me like a deeper issue with jit and scan.

I run the latest jax installed with pip.

```python
import jax.numpy as jnp
import jax

# function
def f(carry,x):    
    def invF(x0,x1):
        rotangle0 = x1[0]
        rot0 = jnp.dot(jnp.stack(
            (jnp.stack((jnp.cos(rotangle0),-jnp.sin(rotangle0))),
             jnp.stack((jnp.sin(rotangle0),jnp.cos(rotangle0))))),
            x0[1:])        
        return jnp.stack([rot0[0],rot0[1]])
    invJF = jax.jacfwd(invF,argnums=0)  
    
    def cross(a, b):
        return jnp.array([
            a[1]*b[2] - a[2]*b[1],
            a[2]*b[0] - a[0]*b[2],
            a[0]*b[1] - a[1]*b[0]])    
 
    def invF2():
        b1 = jnp.array([1.,0.,0.])
        b3 = 0*cross(b1,b1) # try e.g. b3 = 0*b1 or b3 = 0*jax.numpy.cross(b1,b1) instead - it changes the result of xs_lax_scan, but obviously it shouldn't and it doesn't affect the python scan
        Rinvx = jnp.stack((b1,0*b1,0*b3),axis=1)
        return Rinvx[0][0:2]
    
    return jnp.zeros(2),(invJF(jnp.array([1.,0.,0.]),carry)[:,1],invF2())

# jax.lax.scan
_,xs_lax_scan = jax.lax.scan(f,
        jnp.zeros(2),
        jnp.zeros(2))
with jax.disable_jit():
    _,xs_lax_scan_wo_jit = jax.lax.scan(f,
        jnp.zeros(2),
        jnp.zeros(2))

# scan python version (from https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html)
import numpy as np
def python_scan(f, init, xs, length=None):
  if xs is None:
    xs = [None] * length
  carry = init
  ys = []
  for x in xs:
    carry, y = f(carry, x)
    ys.append(y)
  return carry, np.stack(ys)

_,xs_python_scan = python_scan(f,
        jnp.zeros(2),
        jnp.zeros(2))

# results - should be identical, but they are not
print("lax.scan result:     ",xs_lax_scan[0][0])
print("lax.scan without jit:",xs_lax_scan_wo_jit[0][0])
print("python scan result:  ",xs_python_scan[0][0])

# direct function evaluation - matches result of python_scan
print("direct eval of f:    ",f(jnp.zeros(2),None)[1][0])
```

gives

```
lax.scan result:      [0. 0.]
lax.scan without jit: [1. 0.]
python scan result:   [1. 0.]
direct eval of f:     [1. 0.]
```

i.e., the lax.scan gives a different result compared to without jit, direct python version, and direct application of f.

**Comment by mattjj** [comment]
Thanks for the excellent bug report!

By the way, if you want to get 'scan without jit', you can use `with jax.disable_jit(): ...`. That is basically a debug mode which, in addition to making `jit` a no-op, also swaps out control flow constructs with their Python equivalents (e.g. [here's what `scan` does](https://github.com/google/jax/blob/c94f41290cc5c28fba059e172279994107c3103e/jax/_src/lax/control_flow.py#L1252-L1263), basically the same as your `python_scan`.)

On a hunch, I ran this on TPU in colab, and got this result:

```
lax.scan result:      [1. 0.]
lax.scan without jit: [1. 0.]
python scan result:   [1. 0.]
direct eval of f:     [1. 0.]
```

I got the same on GPU. This made me think it's an XLA:CPU bug (which is sometimes an LLVM bug).

However, then I ran it on the latest XLA:CPU version (which we haven't yet released as a jaxlib version), and again got the correct result! So this makes me think this bug has already been fixed, and we just need to release a new jaxlib.

Which jaxlib version are you on? (e.g. `pip show jaxlib`, and maybe do a `pip install --upgrade jaxlib` and try the repro again if you haven't done that recently)

**Comment by stefansommer** [comment]
Thanks for the quick response! I did a clean virtualenv and install of jaxlib when testing it. `python show jaxlib` gives `Version: 0.1.69`. Can I manually install the latest XLA:CPU to test with that?

**Comment by lennonhill** [comment]
The example runs fine since a204f545f9efec28aa01f3c1cf089fa7d50b839c, meaning that this should be fixed with jaxlib 0.1.70.

**Comment by stefansommer** [comment]
I can confirm the issue does is gone with jaxlib 0.1.70. Thank you very much for helping out on this!
